csrc/device_lower/pass/allocation.cpp

```diff
@@ -560,16 +560,24 @@ class AllocationInserter : public kir::ExprMutator {
                                  .build();
       mbarrier->setMemoryType(MemoryType::Shared);
       auto mbarrier_init = IrBuilder::create<kir::MBarrierInit>(
-          mbarrier, expr->container()->oneVal(DataType::UInt32));
+          mbarrier,
+          simplifyExpr(SimplifyingIrBuilder::maybeCastExpr(
+              DataType::UInt32,
+              lower_utils::getNumThreadsInTensorView(
```

Comment:
The idea here is to initialize mbarrier with the number of threads (equal to product of TVs' extends) defined for `LoadStoreOp`'s output. With this change the expectation is that mbarrier will be released not only when transaction is done, by also when defined number of threads reach it.

Is this initialization predicated, so it won't end with multiple threads initializing the same mbarrier object (the same allocation)?

For circular/double buffer functionality I assumed that the requirement for mbarrier to be released is when transaction is done, while it is enough if a single thread reach barrier (weaker requirements). Do you think this should be redesigned?

Reply:
No, this initialization is not predicated, and yes, we will have all threads initializing the same mbarrier. I don't know if this is defined behavior, but we definitely need to revisit this in the future. For this PR, as long as it in practice works, I will leave it like that.

Reply:
> With this change the expectation is that mbarrier will be released not only when transaction is done, by also when defined number of threads reach it.

I think it is an `and` instead of `or` for these two conditions? mbarrier will be release when both the defined number of threads has reached it and the transaction is done.

Reply:
> Do you think this should be redesigned?

Given that the release condition is an `and` instead of an `or`, I don't think it needs to be redesigned?

Reply:
Let's keep implementation as it is, as it seems to be solution that works well for most, if not all, use cases.

I would need to check if there are no issues if for example we set condition: 8 threads + some amount of bytes to be transferred, and before mem transaction is done, more than 8 threads reach this barrier. Phase transition should happened when both conditions are valid, I'm not sure if having value larger than condition won't cause sync issues.